### PR TITLE
Move System.Transactions.Local over to using source-generated COM instead of built-in COM whenever possible.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
This PR moves System.Transactions.Local to use source-generated COM using our analyzers/code-fixes with a few minor adjustments at API boundaries. Majority of the work was done by directly following the analyzers/fixers with minimal additional code added.

There are a few interesting elements to this conversion:

- `ITransactionOutcomeEvents` was not converted to source-generated COM as it is passed to `System.Runtime.InteropServices.ComTypes.IConnectionPoint`, which is provided as a built-in COM-based interface.
   - We can include source-generated-COM definitions of these interfaces, but I didn't want to do that in this PR as this PR is meant to represent a "typical" experience of migration, and copying in those interfaces manually and owning copies of them would not be typical behavior we'd expect.
- `IDtcTransaction` is a public `[ComImport]` interface, so to ensure back-compat, I kept it as that and used `Marshal.Get/SetComObjectData` to provide a clean translation layer between the two worlds (built-in COM externally, source-generated COM internally).
- We needed to disable runtime marshalling due to all of the `Guid` parameters.
- We needed to add one custom marshaller.
- This PR includes https://github.com/dotnet/runtime/pull/87342

### CODE REVIEW

File: src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcInterfaces/IPrepareInfo.cs [hunk 1]
```diff
@@ -2,14 +2,15 @@
...
 
 // https://docs.microsoft.com/previous-versions/windows/desktop/ms686533(v=vs.85)
-[ComImport, Guid("80c7bfd0-87ee-11ce-8081-0080c758527e"), InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
-internal interface IPrepareInfo
+[GeneratedComInterface, Guid("80c7bfd0-87ee-11ce-8081-0080c758527e"), InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
+internal partial interface IPrepareInfo
 {
     void GetPrepareInfoSize(out uint pcbPrepInfo);
 
-    void GetPrepareInfo([MarshalAs(UnmanagedType.LPArray), Out] byte[] pPrepInfo);
```
Comment (jtschuster) [inline src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcInterfaces/IPrepareInfo.cs:14]: Was removing this `Out` intentional? If so, why?

File: src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcInterfaces/ITransactionOutcomeEvents.cs [hunk 2]
```diff
@@ -2,13 +2,13 @@
...
 using System.Runtime.InteropServices;
-using System.Transactions.Oletx;
 
 namespace System.Transactions.DtcProxyShim.DtcInterfaces;
 
+// Keep this type as a ComImport type as it is used with IConnectionPointContainer (and as a result needs to use built-in COM).
 // https://docs.microsoft.com/previous-versions/windows/desktop/ms686465(v=vs.85)
 [ComImport, Guid("3A6AD9E2-23B9-11cf-AD60-00AA00A74CCD"), InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
-internal interface ITransactionOutcomeEvents
+internal partial interface ITransactionOutcomeEvents
```
Comment (jtschuster) [inline src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcInterfaces/ITransactionOutcomeEvents.cs:11]: If we're leaving this as a ComImport, we don't need this to be partial, do we?

File: src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcInterfaces/IPrepareInfo.cs (hunk 1 above)
Comment (jkoritzinsky) [reply src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcInterfaces/IPrepareInfo.cs:14]: Yes, `[Out]` does nothing on arrays of blittable values, so we don't allow it.

**Review by AaronRobinsonMSFT — COMMENTED** [review]
LGTM. Consider me signed off as long as @roji is good with the changes.

File: src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcInterfaces/IPrepareInfo.cs (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcInterfaces/IPrepareInfo.cs:14]: The `[Out]` attribute is incredibly confusing and often causes more trouble than help. @jkoritzinsky or @elinor-fung didn't we right up details on it in the original design doc?

File: src/libraries/System.Transactions.Local/src/System.Transactions.Local.csproj [hunk 3]
```diff
@@ -3,7 +3,7 @@
     <IsPartialFacadeAssembly>true</IsPartialFacadeAssembly>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <TargetFrameworks>$(NetCoreAppCurrent)-windows;$(NetCoreAppCurrent)</TargetFrameworks>
-    <NoWarn>CA1805;IDE0059;CS1591</NoWarn>
+    <NoWarn>CA1805;IDE0059;CS1591;CS3016</NoWarn>
```
Comment (jkoritzinsky) [inline src/libraries/System.Transactions.Local/src/System.Transactions.Local.csproj:6]: NoWarn due to https://github.com/dotnet/roslyn/issues/68526

File: src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/NativeEnums.cs [hunk 4]
```diff
@@ -115,7 +115,7 @@ internal enum OletxTransactionHeuristic : uint
     XACTHEURISTIC_DANGER = 4
 }
 
-internal enum OletxXactStat
+internal enum OletxXactStat : int
```
Comment (roji) [inline src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/NativeEnums.cs:118]: Isn't int the default anyway? Is this about making this more explicit for some reason?

File: src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/Xactopt.cs [hunk 5]
```diff
@@ -17,4 +19,34 @@ internal Xactopt(uint ulTimeout, string szDescription)
...
+            // Usage of Xactopt never passes non-ASCII chars, so we can ignore them.
+            Encoding.ASCII.TryGetBytes(managed.SzDescription, new Span<byte>(native.SzDescription, 40), out _);
+
+            return native;
+        }
+
+        public static unsafe Xactopt ConvertToManaged(XactoptNative unmanaged)
+        {
+            return new(unmanaged.UlTimeout, Encoding.ASCII.GetString(unmanaged.SzDescription, 40));
+        }
```
Comment (roji) [inline src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/Xactopt.cs:50]: nit:

```suggestion
        => new(unmanaged.UlTimeout, Encoding.ASCII.GetString(unmanaged.SzDescription, 40));
```

**Review by roji — APPROVED** [review]
Am no expert on COM and the changes below - definitely fine with them if @AaronRobinsonMSFT signed off and all tests pass, thanks!

File: src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/NativeEnums.cs (hunk 4 above)
Comment (jkoritzinsky) [reply src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/NativeEnums.cs:118]: Yes, this is the default. I added this as some of the COM definitions we’re using MarshalAs for this enum type and explicitly specifying UnmanagedType.I4, so I wanted to be consistent within the project (that specifies uint for cases where UnmanagedType.U4 was used).

File: src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcInterfaces/IPrepareInfo.cs (hunk 1 above)
Comment (jkoritzinsky) [reply src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcInterfaces/IPrepareInfo.cs:14]: I think we did. I’ll make sure on Monday

Comment (jkoritzinsky) [reply src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcInterfaces/IPrepareInfo.cs:14]: Yep, we have it in our compatibility doc right here: https://github.dev/dotnet/runtime/pull/87351

**Comment by jkoritzinsky** [comment]
Failures/timeouts are unrelated.

**Pull request merged**